### ISSUE TITLE
[ascend] tl.cumsum over a boolean (i1) tensor fails lowering — hfusion.cumsum refuses i1 operands

### ISSUE THREAD

**Issue by jianzs** [triton-lang/triton-ascend#2334]
## Summary

On triton-ascend, `tl.cumsum` over a boolean (`i1`) tensor fails compilation: the backend lowering (`hfusion.cumsum`) refuses `i1` operands. The same kernel compiles and produces correct results if the operand is cast to `int32` first.

This blocks upstream vLLM's top-k/top-p sampling kernel (`vllm/v1/sample/ops/topk_topp_triton.py`), which uses `tl.cumsum` over boolean masks.

## Reproducer

```python
import torch, torch_npu, triton, triton.language as tl

@triton.jit
def cumsum_bool_kernel(x_ptr, o_ptr, N: tl.constexpr):
    offs = tl.arange(0, N)
    x = tl.load(x_ptr + offs)
    m = x > 0
    c = tl.cumsum(m.to(tl.int1), axis=0)   # boolean operand
    tl.store(o_ptr + offs, c)

x = torch.randn(64, device="npu:0")
o = torch.zeros(64, device="npu:0", dtype=torch.int32)
cumsum_bool_kernel[(1,)](x, o, 64)
torch.npu.synchronize()
```

Observed (stable):

```
triton.compiler.errors.MLIRCompilationError:
[ConvertLinalgRToBinary] encounters error:
loc(.../cumsum_bool_repro.py:5:0): error: Failed to run BiShengHIR pipeline
loc(callsite(".../triton/language/standard.py":314:60 at ".../cumsum_bool_repro.py":9:18)):
error: 'hfusion.cumsum' op operand #0 must be ranked tensor of bfloat16 type or 16-bit
float or 32-bit float or 8-bit signless integer or 16-bit signless integer or 32-bit
signless integer or 64-bit signless integer values, but got 'tensor<64xi1>'
[ERROR] Failed to run BiShengIR pipeline
```

Contrast — same kernel with an explicit cast **compiles and produces a numerically exact result**:

```python
c = tl.cumsum(m.to(tl.int32), axis=0)   # => PASS, matches torch.cumsum(x > 0)
```

## Expected behavior

Either:
- `tl.cumsum` should accept `i1` operands (upstream Triton lowers them), or
- if the backend intentionally restricts operand types, the frontend should reject it with a clear type error instead of failing deep inside `bishengir-compile` with an MLIR IR dump.

## Environment

- Hardware: Ascend 910 (A3 series)
- CANN: 9.1.0
- Python: 3.12.13
- torch: 2.10.0+cpu; torch_npu: 2.10.0.post4
- triton (imported at runtime): 3.2.0
- triton-ascend: 3.2.2 (`bishengir-compile` under `site-packages/triton/backends/ascend/bishengir/bin/`)

**Comment by hongziqi** [comment]
Thanks for your feedback. We reproduced the failure on the **official Triton-Ascend 3.2.2 release** package. Compilation fails in `bishengir-compile` with:

```text
error: 'hfusion.cumsum' op operand #0 must be ranked tensor of bfloat16 type or 16-bit
float or 32-bit float or 8-bit signless integer or 16-bit signless integer or 32-bit
signless integer or 64-bit signless integer values, but got 'tensor<64xi1>'
[ERROR] Failed to run BiShengIR pipeline
```

The Ascend adapter IR for this kernel is fine. The rejection happens later inside **NPUIR**, which historically did not accept `i1` as a `cumsum` operand type.

## Fix / recommendation

This has been fixed in an updated `bishengir-compile` bundled with the **3.2.2 daily** wheels. The official 3.2.2 release package has not yet picked up that BiShengIR update, so the same repro still fails there.

Please install a **3.2.2 daily** build (2026-09-22) instead of the older release wheel:

- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp310-cp310-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp310-cp310-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp311-cp311-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp312-cp312-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp312-cp312-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp313-cp313-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp313-cp313-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp39-cp39-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl

We re-verified the original boolean `tl.cumsum` repro on that daily package: compilation succeeds.

**Comment by jianzs** [comment]
<img width="1050" height="230" alt="Image" src="https://github.com/user-attachments/assets/f86124c2-a678-499a-847b-2f3b3a1914ce" />

**Comment by jianzs** [comment]
> Thanks for your feedback. We reproduced the failure on the **official Triton-Ascend 3.2.2 release** package. Compilation fails in `bishengir-compile` with:
> 
> ```
> error: 'hfusion.cumsum' op operand #0 must be ranked tensor of bfloat16 type or 16-bit
> float or 32-bit float or 8-bit signless integer or 16-bit signless integer or 32-bit
> signless integer or 64-bit signless integer values, but got 'tensor<64xi1>'
> [ERROR] Failed to run BiShengIR pipeline
> ```
> 
> The Ascend adapter IR for this kernel is fine. The rejection happens later inside **NPUIR**, which historically did not accept `i1` as a `cumsum` operand type.
> 
> ## Fix / recommendation
> This has been fixed in an updated `bishengir-compile` bundled with the **3.2.2 daily** wheels. The official 3.2.2 release package has not yet picked up that BiShengIR update, so the same repro still fails there.
> 
> Please install a **3.2.2 daily** build (2026-09-22) instead of the older release wheel:
> 
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp310-cp310-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp310-cp310-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp311-cp311-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp312-cp312-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp312-cp312-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp313-cp313-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp313-cp313-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp39-cp39-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> 
> We re-verified the original boolean `tl.cumsum` repro on that daily package: compilation succeeds.

Once a release version is out, there shouldn’t be any differences under the same version number. If changes are needed, they should go out as a minor bugfix release. Having a ”daily" release version is a bit odd.

**Comment by hongziqi** [comment]
> > Thanks for your feedback. We reproduced the failure on the **official Triton-Ascend 3.2.2 release** package. Compilation fails in `bishengir-compile` with:
> > ```
> > error: 'hfusion.cumsum' op operand #0 must be ranked tensor of bfloat16 type or 16-bit
> > float or 32-bit float or 8-bit signless integer or 16-bit signless integer or 32-bit
> > signless integer or 64-bit signless integer values, but got 'tensor<64xi1>'
> > [ERROR] Failed to run BiShengIR pipeline
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The Ascend adapter IR for this kernel is fine. The rejection happens later inside **NPUIR**, which historically did not accept `i1` as a `cumsum` operand type.
> > ## Fix / recommendation
> > This has been fixed in an updated `bishengir-compile` bundled with the **3.2.2 daily** wheels. The official 3.2.2 release package has not yet picked up that BiShengIR update, so the same repro still fails there.
> > Please install a **3.2.2 daily** build (2026-09-22) instead of the older release wheel:
> > 
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp310-cp310-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp310-cp310-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp311-cp311-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp312-cp312-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp312-cp312-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp313-cp313-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp313-cp313-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> > * https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend-322/20260922225103/triton_ascend-3.2.2-cp39-cp39-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
> > 
> > We re-verified the original boolean `tl.cumsum` repro on that daily package: compilation succeeds.
> 
> Once a release version is out, there shouldn’t be any differences under the same version number. If changes are needed, they should go out as a minor bugfix release. Having a ”daily" release version is a bit odd.

Version 3.2.2 is still under maintenance, with a small set of bug fixes being merged. We plan to release a new official version 3.2.2.post, so daily wheels are available for this branch. Since the main branch hasn’t had an official release yet, it also provides daily wheels.

**Comment by hongziqi** [comment]
> <img alt="Image" width="1050" height="230" src="https://private-user-images.githubusercontent.com/20160766/657015930-f86124c2-a678-499a-847b-2f3b3a1914ce.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3OTAxNDU4MzAsIm5iZiI6MTc5MDE0NTUzMCwicGF0aCI6Ii8yMDE2MDc2Ni82NTcwMTU5MzAtZjg2MTI0YzItYTY3OC00OTlhLTg0N2ItMmYzYjNhMTkxNGNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA5MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwOTIzVDA2Mzg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyYTBjMWJjZTJhOGYxZDhmNTkxZTlhYWI5Njk2N2ZlNzlhN2VjYWUxY2E2MDk5MzU0M2Y5ODhhNmM2OGEzZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.e2IRtcWYUSGjQRXvtlvhhVeuksFdw2HYihcbNFNhay0">


Thanks for your testing feedback. I will continue to investigate issues #2353 and #2354 later.
